kernel + posix: add k_spin_trylock(), pthread_spin_lock(), etc

cfriedt opened the pull request:
This change adds a new spinlock API, `k_spin_trylock()`, in order to implement `pthread_spin_lock()` and friends. Part of the POSIX Roadmap for LTSv3.

Fixes #59910

Commit:
kernel: spinlock: add k_spin_trylock()

To satisfy the requirements of POSIX spinlocks, we need a common
way to attempt to lock a spinlock that will return an error if
unsuccessful.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: kernel: spinlock: add test for k_spin_trylock()

Add a testcase to exercise two cases:
* when `k_spin_trylock()` fails (lock is busy)
* when `k_spin_trylock()` succeeds (lock is acquired)

We use the same machinery for checking for a recursive mutex
as `k_spin_lock()`.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: pthread: implement pthread spinlocks

pthread_spinlock_t support is part of the base threading
requirement.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: posix: common: add pthread spinlock tests

Exercise the following (which rely mainly on existing Zephyr
utilities)
* `pthread_spin_init()`
* `pthread_spin_destroy()`
* `pthread_spin_lock()`
* `pthread_spin_trylock()`
* `pthread_spin_unlock()`

Additionally, ensure that we do not leak spinlock descriptors.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
doc: services: portability: posix: support for POSIX_SPIN_LOCKS

Ensure `POSIX_SPIN_LOCKS` are documented as part of the
mandatory option requirements according to IEEE Std 1003.1-2017.

Additionally, list the functions implemented according to the
standard.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

evgeniy-paltsev commented on lib/posix/pthread_spinlock.c:
```diff
@@ -0,0 +1,157 @@
+	return l - posix_spinlock_pool;
+#else
+	/* Workaround for struct k_spinlock being zero bytes */
+	__ASSERT_NO_MSG(sizeof(struct k_spinlock))
```
Why runtime assertion? Let's use `BUILD_ASSERT`.

cfriedt replied:
Good idea

cfriedt replied:
Fixed!

ycsin reviewed (dismissed):
LGTM

cfriedt commented:
@evgeniy-paltsev - will you take another look?

andyross commented on include/zephyr/spinlock.h:
```diff
@@ -144,28 +168,50 @@ static ALWAYS_INLINE k_spinlock_key_t k_spin_lock(struct k_spinlock *l)
+	z_spinlock_validate_pre(l);
+#ifdef CONFIG_SMP
+	if (!atomic_cas(&l->locked, 0, 1)) {
+		arch_spin_relax();
```
Best not to put the relax call in here, since there's no loop to relax.  The caller can do that on their own if needed.

cfriedt replied:
Fixed!

andyross commented on include/zephyr/spinlock.h:
```diff
@@ -105,6 +107,28 @@ bool z_spin_lock_mem_coherent(struct k_spinlock *l);
+#ifdef CONFIG_KERNEL_COHERENCE
+	__ASSERT_NO_MSG(z_spin_lock_mem_coherent(l));
+#endif
+#endif
```
Not really the fault of this patch, but this clearly some #if'ery refactoring.  Testing the kconfig inside a function named "validate" seems silly.  Likewise we should just define z_spin_lock_mem_coherent() to a nop if !COHERENCE.

andyross replied:
No, it was just a drive-by comment.  But feel free if you want to. :)

cfriedt replied:
I would, but meetings (on no-meeting day no less!)

peter-mitsis commented on include/zephyr/spinlock.h:
```diff
@@ -144,28 +168,49 @@ static ALWAYS_INLINE k_spinlock_key_t k_spin_lock(struct k_spinlock *l)
+/**
+ * @brief Attempt to lock a spinlock
+ *
+ * This routine makes one attempt to lock @p l. If it is successful, then
+ * it will store the key into @p k.
```
Minor question -- should this be using @a instead of @p ?

peter-mitsis reviewed (dismissed):
Looks good to me.

ycsin reviewed (dismissed):
re-stamp

Commit:
checkpatch: exclude lib/posix/*.c from negative errno rule

The `USE_NEGATIVE_ERRNO` rule consistently generates false
positives when working with certain areas of POSIX.

It makes sense to disable this check for the POSIX API
rather than requiring merge superpowers constantly. That
way, we can merge as per usual after sufficient approvals
rather than waiting for someone with merge superpowers to
override / manually merge.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

xyzzy42 commented:
There is some opportunity to increase compiler optimization by providing the public pthread methods as inline wrappers that take a `pthread_spinlock_t*` argument and passes it to an internal function, but first convert the argument to a `pthread_spinlock_t` or `bit`.

The check that `lock` is not NULL can likely be resolved by the compiler at compile time.   It's possible to avoid even allocating space for `lock` and have it live in a register (would need a somewhat limited scope for the lock to be defined over to actually be done).  `*lock` needs to be loaded from RAM on each call, since the compiler can not determine nothing changes it.  But if an inline function (or LTO!) is used, the compiler can see that nothing changes the value and avoid the instructions to reload it.

cfriedt commented:
> There is some opportunity to increase compiler optimization by providing the public pthread methods as inline wrappers that take a `pthread_spinlock_t*` argument and passes it to an internal function, but first convert the argument to a `pthread_spinlock_t` or `bit`.

There is always room for optimization but that can be done on an as-needed basis at a later date with evidence / measurements. Additionally, having something that works and meets the spec is more important, and I would avoid inlining anything that isn't specifically mentioned that it could possibly be an inline function or macro.

> The check that `lock` is not NULL can likely be resolved by the compiler at compile time. It's possible to avoid even 

Technically, that check isn't even required by the standard and it represents undefined behaviour. I was thinking about making it an `__ASSERT_NO_MSG()` since that can easily be disabled.

> allocating space for `lock` and have it live in a register (would need a somewhat limited scope for the lock to be defined over to actually be done). `*lock` needs to be loaded from RAM on each call, since the compiler can not determine nothing changes it. But if an inline function (or LTO!) is used, the compiler can see that nothing changes the value and avoid the instructions to reload it.

Yep - keeping `lock` in a register would have benefits, but it also isn't clear to me how to return multiple values in registers. Additionally, there is no "invalid lock" value as part of the `k_spinlock` API. We can leave that to future optimization (ideally before the next release).

andyross reviewed (approved):
Refresh +1

xyzzy42 commented:
> Technically, that check isn't even required by the standard and it represents undefined behaviour. I was thinking about making it an `__ASSERT_NO_MSG()` since that can easily be disabled.

The same also for both checks for use of an uninitialized spinlock, undefined behavior in POSIX, but certainly a nice debugging aid.  The second check makes all spinlocks contend for access to the single spinlock protecting `posix_spinlock_bitarray`.  I suppose scaling probably isn't much of a concern.  The flash sharing caused by having all spinlocks be adjacent in memory will not help scaling either.
